feat: add C implementation for `stats/base/dists/lognormal/pdf`

NishchayRajput opened the pull request:
---
Resolves #3755.

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for @stdlib/stats/base/dists/lognormal/pdf along with relevant tests, benchmarks and examples.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3755 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

NishchayRajput committed:
feat: add C implementation for @stdlib/stats/base/dists/lognormal/pdf

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

AadishJ commented:
You need to add 3 lines in the package.json
![image](https://github.com/user-attachments/assets/ca84b4af-dfb4-46e3-9a01-12a2efbdd810)

NishchayRajput committed:
fix: updated package.json

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

NishchayRajput committed:
fix: updated package.json

NishchayRajput commented:
> You need to add 3 lines in the package.json ![image](https://private-user-images.githubusercontent.com/71633935/400154910-ca84b4af-dfb4-46e3-9a01-12a2efbdd810.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYwMTM3MDAsIm5iZiI6MTczNjAxMzQwMCwicGF0aCI6Ii83MTYzMzkzNS80MDAxNTQ5MTAtY2E4NGI0YWYtZGZiNC00NmUzLTlhMDEtMTJhMmVmYmRkODEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA0VDE3NTY0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIwMTRjOTk1M2NkODllNDViMjY1MzcxNzQ1NjIyODg3YjFlMjQyODEwZjk2NDMyMDFiYTY0ZWFlNmRmNDI2ZTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iIdSOIPNhvhK96KTYOEqHDmVCbXSqHja6D6FFjnzCng)

done

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/README.md:
````diff
@@ -138,6 +138,105 @@ for ( i = 0; i < 10; i++ ) {
+### Usage
+
+```c
+#include "stdlib/stats/base/dists/lognorma/pdf.h"
````
typo

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/README.md:
````diff
@@ -138,6 +138,105 @@ for ( i = 0; i < 10; i++ ) {
+#include "stdlib/stats/base/dists/lognorma/pdf.h"
+```
+
+#### stdlib_base_dists_lognormal_pdf( x, a, b, c )
````
wrong input parameters

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/README.md:
```diff
@@ -138,6 +138,105 @@ for ( i = 0; i < 10; i++ ) {
+}
+
+int main( void ) {
+    double x;
+    double sigma;
+    double mu;
+    double y;
+    int i;
```
order

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,74 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2024 The Stdlib Authors.
```
update copyright year

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/benchmark/c/Makefile:
```diff
@@ -0,0 +1,146 @@
+#/
+# @license Apache-2.0
+#
+# Copyright (c) 2024 The Stdlib Authors.
```
update copyright year

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,143 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2024 The Stdlib Authors.
```
update copyright year

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,143 @@
+*/
+static double benchmark( void ) {
+	double elapsed;
+    double x[ 100 ];
```
indentation

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,143 @@
+	int i;
+
+	for ( i = 0; i < 100; i++ ) {
+        x[ i ] = random_uniform( 0.0, 100.0 ) - 100.0;
```
indentation

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/binding.gyp:
```diff
@@ -0,0 +1,170 @@
+# @license Apache-2.0
+#
+# Copyright (c) 2024 The Stdlib Authors.
```
update copyright year

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/examples/c/Makefile:
```diff
@@ -0,0 +1,146 @@
+#/
+# @license Apache-2.0
+#
+# Copyright (c) 2024 The Stdlib Authors.
```
update copyright year

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/examples/c/example.c:
```diff
@@ -0,0 +1,42 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2024 The Stdlib Authors.
```
update copyright year

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/examples/c/example.c:
```diff
@@ -0,0 +1,42 @@
+}
+
+int main( void ) {
+    double x;
```
indentation

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/include.gypi:
```diff
@@ -0,0 +1,53 @@
+# @license Apache-2.0
+#
+# Copyright (c) 2024 The Stdlib Authors.
```
copyright year

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/include/stdlib/stats/base/dists/lognormal/pdf.h:
```diff
@@ -0,0 +1,38 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2024 The Stdlib Authors.
```
update copyright year

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/src/Makefile:
```diff
@@ -0,0 +1,70 @@
+#/
+# @license Apache-2.0
+#
+# Copyright (c) 2024 The Stdlib Authors.
```
update copyright year

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/src/addon.c:
```diff
@@ -0,0 +1,23 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2024 The Stdlib Authors.
```
update copyright year

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/src/main.c:
```diff
@@ -0,0 +1,63 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2024 The Stdlib Authors.
```
update copyright year

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/src/main.c:
```diff
@@ -0,0 +1,63 @@
+* // returns ~0.157
+*/
+double stdlib_base_dists_lognormal_pdf(const double x, const double mu, const double sigma ) {
+	double s2;
```
indentation

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/src/main.c:
```diff
@@ -0,0 +1,63 @@
+    double exp_value;
+    double lnx_u;
+    double z;
+    double y;
```
order

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/test/test.native.js:
```diff
@@ -0,0 +1,190 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2024 The Stdlib Authors.
```
update copyright year

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/examples/c/example.c:
```diff
@@ -0,0 +1,42 @@
+    double x;
+	double sigma;
+	double mu;
+	double y;
```
sequence

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/examples/c/example.c:
```diff
@@ -0,0 +1,42 @@
+	int i;
+
+	for ( i = 0; i < 25; i++ ) {
+        x = random_uniform( 0.0, 10.0 );
```
indentation

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/lib/native.js:
```diff
@@ -0,0 +1,77 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2024 The Stdlib Authors.
```
update copyright years

NishchayRajput committed:
fix: refactor random function

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

NishchayRajput commented:
@Planeshifter Please review my PR. If there is some issue then I can fix in other PRs as well.

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/benchmark/benchmark.js:
```diff
@@ -32,17 +33,25 @@ var pdf = require( './../lib' );
+	mu = new Float64Array( len );
+	sigma = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = ( randu() * 100.0 ) - 100.0;
+		mu[ i ] = ( randu() * 100.0 ) - 50.0;
+		sigma[ i ] = ( randu() * 20.0 ) + EPS;
```
use uniform form `@stdlib/random/base/uniform`

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/include/stdlib/stats/base/dists/lognormal/pdf.h:
```diff
@@ -0,0 +1,38 @@
+#endif
+
+/**
+* Returns the probability density function (PDF) for a lognormal distribution with location `mu` and scale `sigma`.
```
Returns => `Evaluates`

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/src/main.c:
```diff
@@ -0,0 +1,63 @@
+double stdlib_base_dists_lognormal_pdf( const double x, const double mu, const double sigma ) {
+    double exp_value;
+    double lnx_u;
+	double s2;
```
indentation

NishchayRajput commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/src/main.c:
```diff
@@ -0,0 +1,63 @@
+    double z;
+    double y;
+	if (
+        stdlib_base_is_nan( x ) ||
+		stdlib_base_is_nan( mu ) ||
+		stdlib_base_is_nan( sigma ) ||
+		sigma <= 0.0
```
indentation

Planeshifter committed:
fix: correct comments, function name, and use stdlib PI constant in C implementation for `stats/base/dists/lognormal/pdf`

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
fix: update documentation and examples to use proper uniform distribution ranges

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
fix: add missing EPS constant dependency to examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
fix: correct indentation in C files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
fix: use Greek symbols in C printf format strings

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/src/addon.c:
```diff
@@ -0,0 +1,23 @@
+#include "stdlib/stats/base/dists/lognormal/pdf.h"
+#include "stdlib/math/base/napi/ternary.h"
+
+// cppcheck-suppress shadowFunction
+STDLIB_MATH_BASE_NAPI_MODULE_DDD_D( stdlib_base_dists_lognormal_pdf )
```
```suggestion
STDLIB_MATH_BASE_NAPI_MODULE_DDD_D( stdlib_base_dists_lognormal_pdf )

```

Planeshifter committed:
style: fix indentation

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/lognormal/pdf/lib/native.js:
```diff
@@ -0,0 +1,77 @@
+// MAIN //
+
+/**
+* Returns the pdf of a lognormal distribution with location parameter `mu` and scale parameter `sigma` at a value `x`.
```
```suggestion
* Evaluates the probability density function (PDF) for a lognormal distribution with location `mu` and scale `sigma`.
```

Planeshifter committed:
docs: update description

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
docs: fix section tags

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter reviewed (dismissed):
Should be ready to land!

Planeshifter committed:
chore: remove cppcheck directive

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
Merge branch 'develop' into nishchay-base-lognormal-pdf

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
style: remove extra space

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
